File: investigation-backend/src/api/routes/attachments.py
from uuid import UUID

from fastapi import APIRouter, Depends, File, HTTPException, UploadFile, status
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from src.api.deps import get_current_user
from src.core.database import get_db
from src.models.attachment import Attachment
from src.models.task import Task
from src.models.user import User
from src.schemas.attachment import AttachmentCreate, AttachmentOut
from src.services.storage import StorageError, get_storage

router = APIRouter(prefix="/tasks/{task_id}/attachments")


async def _ensure_task(
    task_id: UUID, db: AsyncSession, current_user: User
) -> None:
    result = await db.execute(
        select(Task.id).where(Task.id == task_id, Task.user_id == current_user.id)
    )
    if result.scalar_one_or_none() is None:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Task not found")


@router.get("/", response_model=list[AttachmentOut])
async def list_attachments(
    task_id: UUID,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
) -> list[Attachment]:
    await _ensure_task(task_id, db, current_user)
    result = await db.execute(
        select(Attachment)
        .where(Attachment.task_id == task_id, Attachment.user_id == current_user.id)
        .order_by(Attachment.created_at.desc())
    )
    return result.scalars().all()


@router.post("/", response_model=AttachmentOut, status_code=status.HTTP_201_CREATED)
async def create_attachment(
    task_id: UUID,
    payload: AttachmentCreate,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
) -> Attachment:
    await _ensure_task(task_id, db, current_user)

    attachment = Attachment(
        task_id=task_id,
        user_id=current_user.id,
        file_name=payload.file_name,
        file_type=payload.file_type,
        file_size=payload.file_size,
        cos_url=payload.cos_url,
        cos_key=payload.cos_key,
        thumbnail_url=payload.thumbnail_url,
    )
    db.add(attachment)
    await db.commit()
    await db.refresh(attachment)
    return attachment


@router.post("/upload", response_model=AttachmentOut, status_code=status.HTTP_201_CREATED)
async def upload_attachment(
    task_id: UUID,
    file: UploadFile = File(...),
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
) -> Attachment:
    await _ensure_task(task_id, db, current_user)
    storage = get_storage()

    try:
        stored = await storage.save_upload(file, current_user.id, task_id)
    except StorageError as exc:
        raise HTTPException(status_code=exc.status_code, detail=str(exc)) from exc

    attachment = Attachment(
        task_id=task_id,
        user_id=current_user.id,
        file_name=stored.file_name,
        file_type=stored.content_type,
        file_size=stored.size,
        cos_url=stored.url,
        cos_key=stored.key,
        thumbnail_url=None,
    )
    db.add(attachment)
    await db.commit()
    await db.refresh(attachment)
    return attachment


@router.delete("/{attachment_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_attachment(
    task_id: UUID,
    attachment_id: UUID,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
) -> None:
    await _ensure_task(task_id, db, current_user)
    result = await db.execute(
        select(Attachment).where(
            Attachment.id == attachment_id,
            Attachment.task_id == task_id,
            Attachment.user_id == current_user.id,
        )
    )
    attachment = result.scalar_one_or_none()
    if not attachment:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Attachment not found")

    await db.delete(attachment)
    await db.commit()
    return None

File: investigation-admin/src/api/attachments.ts
import { apiRequest } from "./client"

const API_BASE_URL = import.meta.env.VITE_API_BASE_URL || "http://127.0.0.1:8000/api/v1"
const TOKEN_KEY = "admin_token"

export type AttachmentItem = {
  id: string
  task_id: string
  user_id: string
  file_name: string
  file_type: string
  file_size: number
  cos_url: string
  cos_key: string
  thumbnail_url?: string | null
  created_at: string
}

export function fetchAttachments(taskId: string): Promise<AttachmentItem[]> {
  return apiRequest<AttachmentItem[]>(`/tasks/${taskId}/attachments/`)
}

export async function uploadAttachment(taskId: string, file: File): Promise<AttachmentItem> {
  const token = localStorage.getItem(TOKEN_KEY)
  const formData = new FormData()
  formData.append("file", file)

  const response = await fetch(`${API_BASE_URL}/tasks/${taskId}/attachments/upload`, {
    method: "POST",
    headers: token ? { Authorization: `Bearer ${token}` } : undefined,
    body: formData
  })

  if (!response.ok) {
    const payload = await response.json().catch(() => ({}))
    const message = payload?.detail || `Upload failed (${response.status})`
    throw new Error(message)
  }

  return response.json() as Promise<AttachmentItem>
}

export function deleteAttachment(taskId: string, attachmentId: string): Promise<void> {
  return apiRequest<void>(`/tasks/${taskId}/attachments/${attachmentId}`, { method: "DELETE" })
}
